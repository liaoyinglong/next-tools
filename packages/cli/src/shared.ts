import cac from "cac";
import debug from "debug";

export const cli = cac("qwer");

export const createLogger = (name: string) => {
  return debug(`${cli.name}:${name}`);
};
