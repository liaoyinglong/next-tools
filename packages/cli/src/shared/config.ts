import JoyCon from "joycon";
import { resolveSheetId } from "./resolveSheetId";

const joycon = new JoyCon();

export interface Config {
  // 语言文件存放目录，默认为项目根目录的 "./i18n"
  i18nDir?: string;
  // 语言文件名，默认为 "{locale}.i18n.json"，其中 {locale} 会被替换为 locales 中的语言
  i18nFileName?: string;
  // 默认语言，默认为 "zh"，影响到提出出来的文案的默认值
  defaultLocale?: string;
  /**
   * 可以是 id 或者 url，如果是 url 则会自动解析出 id
   * id:  1C9-Dol3oO20W9_FhiVlxNsDhOaaejJIgAZYRkonGmfk
   * url: https://docs.google.com/spreadsheets/d/1C9-Dol3oO20W9_FhiVlxNsDhOaaejJIgAZYRkonGmfk/edit#gid=1740568548
   */
  sheetId?: string;
  // google sheets 工作表名称
  sheetRange?: string;
  /**
   * 对应 语言、i18nKey 所在的列标识符 如(A,B,C...)
   * 传入时将忽略默认的配置
   */
  position?: {
    key: string;
    [key: string]: string;
  };

  /**
   * 从 Google Sheets 哪一行开始解析
   * 默认从第2行开始，模板文件第一行是表头
   * @default 2
   */
  parseStartIndex?: number;
}
export interface InternalConfig extends Config {
  // 默认是 命令运行的目录，一般是项目根目录
  cwd?: string;
  // 支持的语言列表，从 position 中解析出来
  locales?: string[];
}

export function defineConfig(config: Config | Config[]) {
  return config;
}

export const configName = "dune.config.js";

export async function getConfig(): Promise<InternalConfig[]> {
  const res = await joycon.load([configName]);

  const defaultConfig = {
    i18nDir: "./i18n",
    i18nFileName: "{locale}.i18n.json",
    cwd: process.cwd(),
    position: {
      key: "B",
      zh: "C",
      en: "D",
      in: "E",
    },
    parseStartIndex: 2,
  };

  if (res.data) {
    const config: InternalConfig[] = Array.isArray(res.data)
      ? res.data
      : [res.data];
    return config.map((item) => {
      item = Object.assign({}, defaultConfig, item);
      item.locales = Object.keys(item.position).filter(
        (item) => item !== "key"
      );

      item.sheetId = resolveSheetId(item.sheetId);

      return item;
    });
  }

  return [defaultConfig];
}
