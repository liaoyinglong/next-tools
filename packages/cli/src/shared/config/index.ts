import JoyCon from "joycon";
import { Config } from "./types";
import { normalizeConfig } from "./normalizeConfig";

export * from "./types";

const joycon = new JoyCon();

export function defineConfig<T extends Config = Config>(c: T) {
  return c;
}

export const configName = "dune.config.js";

export async function getConfig(): Promise<Config> {
  const res = await joycon.load([configName]);
  return normalizeConfig(res.data ?? {});
}
