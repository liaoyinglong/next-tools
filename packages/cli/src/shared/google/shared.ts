import connect from "connect";
import open from "open";
import { URL } from "url";
import { createLogger } from "../index";

const log = createLogger("google:shared");

function openBrowser(url: string) {
  log.info(`如未自动打开浏览器，可访问以下地址：${url}`);
  open(url, { app: { name: open.apps.chrome } });
}

export function openAndWaitReturnQuery(
  toOpenUrl: string,
  port: number,
  queryKey: string
) {
  return new Promise<string>((resolve) => {
    openBrowser(toOpenUrl);

    const app = connect();
    app.use((req, res, next) => {
      const parsedUrl = new URL(req.url, `http://localhost:${port}`);
      const target = parsedUrl.searchParams.get(queryKey);
      if (target) {
        res.writeHead(200, {
          "content-type": "text/html;charset=utf8",
        });
        res.end(
          `<div>当前窗口可直接关闭</div>
                  <script>
                    window.open("", "_self", "");
                    window.close();
                  </script>`
        );
        server.close();
        server.unref();
        resolve(target);
      }
      next();
    });

    const server = app.listen(port);
  });
}
