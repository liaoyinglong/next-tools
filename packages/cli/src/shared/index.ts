import cac from "cac";
import debug from "debug";
import path from "path";
import fs from "fs-extra";

export const cli = cac("dune");

export const createLogger = (name: string) => {
  const info = debug(`${cli.name}:${name}`);
  return {
    ...info,
    info,
    error: info.extend("error"),
  };
};

debug.enable(`${cli.name}:*`);

/**
 * same with chacheDir
 * homeDir();
 * //=> '~/name'
 *
 * homeDir('bar.js')
 * //=> '~/name/bar.js'
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const homeConfigDir = (() => {
  const isWin = process.platform === "win32";
  const USER_HOME_DIR = isWin ? process.env.USERPROFILE : process.env.HOME;
  const configDir = path.join(USER_HOME_DIR, ".config", cli.name);

  if (!fs.existsSync(configDir)) {
    fs.mkdirSync(configDir);
  }
  return homeConfigDir;

  function homeConfigDir(p?: string): string {
    if (!p) {
      return configDir;
    }
    return path.join(configDir, p);
  }
})();
