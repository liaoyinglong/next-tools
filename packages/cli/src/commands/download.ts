import pMap from "p-map";
import { createLogger } from "../shared";
import { getConfig } from "../shared/config";
import { I18nData } from "../shared/i18nData";
import { resolveSheetData } from "../shared/resolveSheetData";
import { promptI18nConfigEnable } from "../shared/promptConfigEnable";

const log = createLogger("generate");

export async function download() {
  const config = await getConfig();

  let i18nConfigs = await promptI18nConfigEnable(config.i18n);

  let i = 0;
  for (const configItem of i18nConfigs) {
    i++;
    if (!configItem.sheetId || !configItem.sheetRange) {
      log.error(`config.${i} sheetId 或 sheetRange 为空，请检查`);
      continue;
    }
    const sheetData = await resolveSheetData(configItem);

    const statistics = await pMap(
      configItem.locales ?? [],
      async (locale, index) => {
        const i18nData = new I18nData(locale, configItem);
        await i18nData.updateFromSheetData(sheetData.getLangDataJSON(locale));
        await i18nData.saveToDisk();
        return await i18nData.statistic();
      }
    );
    I18nData.printStatistic(`${configItem.sheetRange} 生成结果: `, statistics);
  }
}
