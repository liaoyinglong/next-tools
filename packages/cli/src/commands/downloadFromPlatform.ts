import axios from "axios";
import pMap from "p-map";
import { createLogger } from "../shared";
import { getConfig } from "../shared/config";
import { I18nData } from "../shared/i18nData";

const logger = createLogger("downloadFromPlatform");

/**
 * 从 翻译平台下载翻译文件
 */
export async function downloadFromPlatform() {
  const config = await getConfig();
  // 只有配置了 translatePlatform 才会下载
  // TODO: 暂时只支持一个，后续支持多个
  const i18nConfig = config.i18n?.[0];
  if (!i18nConfig) {
    logger.error("i18n 配置为空");
    return;
  }
  const { translatePlatform, locales } = i18nConfig;
  if (!translatePlatform?.enable) {
    logger.info("translatePlatform.enable 为 false，不会下载");
    return;
  }

  const res = await axios({
    baseURL: translatePlatform.url,
    url: "/v1/dune-i18n/public/findAll",
    params: {
      projectName: translatePlatform.project,
    },
  }).catch((e) => {
    logger.error("获取翻译平台数据失败: %o", e);
    logger.info("将使用默认数据生成 json 文件");
    return { data: {} };
  });

  const data = res.data.data;
  if (!data) {
    logger.error("获取翻译平台数据失败: %o", res.data);
    logger.info("将使用默认数据生成 json 文件");
  }

  await pMap(locales ?? [], async (locale) => {
    const i18nData = new I18nData(locale, i18nConfig);
    await i18nData.updateFromSheetData(data?.[locale] ?? {});
    await i18nData.saveToDisk();
    return i18nData;
  });
}
