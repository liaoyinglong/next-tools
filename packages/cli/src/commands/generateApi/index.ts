import SwaggerParser from "@apidevtools/swagger-parser";
import fs from "fs-extra";
import { compile } from "json-schema-to-typescript";
import _ from "lodash";
import { OpenAPIV3 } from "openapi-types";
import * as os from "os";
import pMap from "p-map";
import path from "path";
import { createLogger } from "../../shared";
import { ApiConfig, getConfig } from "../../shared/config";
import { formatFile } from "../../shared/formatFile";
import { promptApiConfigEnable } from "../../shared/promptConfigEnable";

const log = createLogger("generateApi");

export async function generateApi() {
  const config = await getConfig();
  const apiConfigs = await promptApiConfigEnable(config.api);

  for (const apiConfig of apiConfigs) {
    log.info(`清除旧的api文件: ${apiConfig.output}`);
    await fs.emptydir(apiConfig.output!);
  }

  for (const apiConfig of apiConfigs) {
    log.info("开始解析 %s", apiConfig.swaggerJSONPath);
    const parsed = (await SwaggerParser.dereference(
      apiConfig.swaggerJSONPath
    )) as OpenAPIV3.Document;

    await pMap(Object.entries(parsed.paths), async ([url, pathItemObject]) => {
      log.info("开始生成 %s", url);
      if (pathItemObject) {
        await pMap(
          ["get", "put", "post", "delete", "patch"],
          async (method) => {
            const operationObject = pathItemObject[method];
            if (operationObject) {
              const code = await generateApiRequestCode({
                url: url,
                method: method,
                operationObject: operationObject,
                apiConfig,
              });
              const outputPath = path
                .join(
                  apiConfig.output!,
                  url,
                  apiConfig.enableTs ? `${method}.ts` : `${method}.js`
                )
                .replace(/:/g, "_");
              await fs.ensureFile(outputPath);
              await fs.writeFile(outputPath, code);
            }
          }
        );
      }
    });

    if (apiConfig.format) {
      await formatFile(apiConfig.output!);
    }
  }

  log.info("generateApi Done");
}

/**
 * 根据请求方法，生成请求代码
 */
export async function generateApiRequestCode(options: {
  url: string;
  method: string;
  operationObject: OpenAPIV3.OperationObject;
  apiConfig: ApiConfig;
}): Promise<string> {
  const { method, operationObject, apiConfig } = options;

  const url = (() => {
    if (typeof apiConfig.urlTransformer === "string") {
      return `${apiConfig.urlTransformer}${options.url}`;
    }
    if (typeof apiConfig.urlTransformer === "function") {
      return apiConfig.urlTransformer(options.url);
    }
    return options.url;
  })();

  const seeUrl = apiConfig.swaggerUiUrl
    ? `${apiConfig.swaggerUiUrl}#/${operationObject.tags?.join("/") ?? ""}/${operationObject.operationId
    }`
    : "";

  // 生成的请求构造器的名称，需要使用原始url
  let requestBuilderName = _.camelCase(`${options.url}_${method}_api`);

  // 判断是否有效的js变量
  if (!/^[a-zA-Z_$][0-9a-zA-Z_$]*$/.test(requestBuilderName)) {
    // 不是有效的js变量，使用 _ + 数字
    requestBuilderName = `_${requestBuilderName}`;
  }

  //#region url上参数 例如 /api/v1/users/{userId}
  const urlPathParams = getUrlPathParams(
    (operationObject.parameters as never) ?? []
  );
  const urlPathParamsCode = urlPathParams.length
    ? `urlPathParams: ${JSON.stringify(urlPathParams)},`
    : "";
  //#endregion

  let code: string[] = [
    "// 这个文件由 @dune2/cli 自动生成，不要手动修改，否则会被覆盖",
    apiConfig.RequestBuilderImportPath!,
    apiConfig.requestFnImportPath!,
    apiConfig.queryClientImportPath!,

    "/**",
    `  * ${operationObject.summary}`,
    `  * @tags ${operationObject.tags?.join(",")}`,
    seeUrl && `  * @see ${seeUrl}`,
    `  */`,
  ].filter(Boolean);

  // builder 代码
  let builderCode = `\
export const ${requestBuilderName} = new RequestBuilder({
  url: '${url}',
  method: '${method}',
  ${apiConfig.requestFnImportPath ? "requestFn," : ""}
  ${urlPathParamsCode}
  ${apiConfig.queryClientImportPath ? "queryClient," : ""}
});`
    // 移除空行
    .replace(/,\n\s*}/, "," + os.EOL + "}");
  if (apiConfig.enableTs) {
    builderCode = builderCode.replace(
      "new RequestBuilder(",
      `new RequestBuilder<${requestBuilderName}.Req, ${requestBuilderName}.Res>(`
    );
  }
  code.push(builderCode);

  if (apiConfig.enableTs) {
    // post和put请求需要生成表单fieldsMap
    const isGenrateFieldsMap = [OpenAPIV3.HttpMethods.POST, OpenAPIV3.HttpMethods.PUT].includes(method as OpenAPIV3.HttpMethods);
    // 请求参数类型
    // 响应参数类型
    const [requestParamsTypeCode, responseParamsTypeCode] = await Promise.all([
      compileRequestParams(operationObject, isGenrateFieldsMap),
      compileResponseParams(operationObject, apiConfig),
    ]);

    code.push(`
export namespace ${requestBuilderName} {
 ${requestParamsTypeCode.code}
 
 ${responseParamsTypeCode}
};`);

    if (isGenrateFieldsMap) {
      // 生成表单fieldsMap
      code.push(`
export const ${requestBuilderName}FieldsMap = ${JSON.stringify(requestParamsTypeCode.fieldsMap)}
`)
    }

  }

  return code.join(os.EOL);
}

function getUrlPathParams(parameters: OpenAPIV3.ParameterObject[]) {
  return parameters
    .filter((item) => item.in === "path")
    .map((item) => item.name);
}

/**
 * 这个方法还不完善，只能处理简单的请求参数
 */
async function compileRequestParams(
  operationObject: OpenAPIV3.OperationObject,
  generateFieldsMap = false
) {
  let schema;
  if (operationObject.requestBody) {
    //TODO: 这里也需要处理其他类型
    schema = (operationObject.requestBody as OpenAPIV3.RequestBodyObject)
      .content["application/json"].schema;
  } else if (operationObject.parameters) {
    const extraProperties = {};
    const parameters: OpenAPIV3.ParameterObject[] = [];
    (operationObject.parameters as OpenAPIV3.ParameterObject[]).forEach(
      (item) => {
        if (!["query", "path"].includes(item.in)) {
          return;
        }
        if (_.get(item.schema, "type") === "object") {
          // swagger get 请求上 有些参数是 object 类型 应该拍平
          _.assign(extraProperties, _.get(item.schema, "properties", {}));
        } else {
          parameters.push(item);
        }
      }
    );

    // 必填参数中忽略 分页相关的参数
    const required = parameters
      .filter(
        (p) => p.required && !["pageNum", "pageSize", "count"].includes(p.name)
      )
      .map((p) => p.name);
    const properties = Object.fromEntries(
      parameters
        // 后端 swagger 可能出现没有 schema 的情况，这里过滤掉
        .filter((p) => !!p.schema)
        .map((p) => {
          //TODO: 这里也要处理 ref 类型
          const schema = p.schema as OpenAPIV3.SchemaObject;
          return [
            p.name,
            {
              type: schema.type,
              description: p.description,
              // enum: schema.enum ?? [],
            },
          ];
        })
    );
    schema = {
      required,
      type: "object",
      properties: { ...properties, ...extraProperties },
    };
  }

  let code = "";
  const fieldsMap: Record<string, string> = {};
  if (schema) {
    try {
      code = await compile(schema, "Req", {
        bannerComment: "",
        ignoreMinAndMaxItems: !!1,
        additionalProperties: false,
        // format: false,
      });

      // 判断接口是否是分页请求
      const isPageSearchRequest = (data: any) => {
        // 入参有这些字段就认为是分页查询接口
        const keys = ['pageNum', 'pageSize', 'count'];
        return _.get(data, 'type') === 'object' &&
          _.get(data, 'required', []).filter(item => keys.includes(item)).length === keys.length
      }

      if (generateFieldsMap) {
        // 分页查询接口取params字段
        const params = isPageSearchRequest(schema) ? _.get(schema, 'properties.params.properties', {}) : schema.properties;
        _.forEach(params, (_, field) => {
          fieldsMap[field] = field
        })
      }
    } catch (e) {
      log.error("生成请求参数类型失败，请检查 %o", {
        summary: operationObject.summary,
        message: e.message,
      });
    }
  }

  return {
    code: code || "export type Req = any;",
    fieldsMap
  }
}

async function compileResponseParams(
  operationObject: OpenAPIV3.OperationObject,
  apiConfig: ApiConfig
) {
  const temp = operationObject.responses["200"] as OpenAPIV3.ResponseObject;
  let code = "";
  if (temp?.content) {
    // FIXME: 可能需要处理其他的content类型
    const temp2 = temp.content["application/json"] || temp.content["*/*"];
    const schema = temp2.schema as OpenAPIV3.SchemaObject;
    let data = apiConfig.responseSchemaTransformer!(schema);
    if (data) {
      // 这里需要深度clone的原因是：
      // 解析出来的scheme会尽可能的被复用，导致影响到下次解析
      data = _.cloneDeep(data);
      markCircularToRef(data);
      try {
        code = await compile(data, "Res", {
          bannerComment: "",
          ignoreMinAndMaxItems: !!1,
          additionalProperties: false,
          // format: false,
        });

        // 通过响应数据判断是否是分页查询接口
        const isPageSearchResponse = (data: any) => {
          // 有result字段且为数组，就认为是分页查询接口
          return _.get(data, 'type') === 'object' && _.get(data, 'properties.result.type') === 'array'
        }

        // 新增后端分页查询返回的数据类型
        if (isPageSearchResponse(data)) {
          code += `${os.EOL}export type ResultItem = Res['result'][0]`
        }
      } catch (e) {
        log.error("转换响应参数类型失败, 请检查 %o", {
          summary: operationObject.summary,
          error: e.message,
        });
      }
    } else {
      log.error("responseSchemaTransformer 返回值为空, 请检查");
    }
  }

  return code ? code : "export type Res = any;";
}

export function markCircularToRef(
  obj,
  parentMark = "#",
  map = new Map([[obj, parentMark]]),
  set = new Set([obj])
) {
  Object.keys(obj).forEach((key) => {
    const value = obj[key];
    if (typeof value === "object" && !Array.isArray(value)) {
      if (set.has(value)) {
        obj[key] = { $ref: map.get(value) };
        return;
      }
      const tempMark = parentMark + "/" + key;
      set.add(value);
      map.set(value, tempMark);
      markCircularToRef(value, tempMark, map, set);
    }
  });
  return map;
}
