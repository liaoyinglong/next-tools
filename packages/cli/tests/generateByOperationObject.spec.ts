import { describe, it, expect } from "vitest";
import { generateApiRequestCode } from "../src/commands/generateApi";
import { apiConfigNormalizer } from "../src/shared/config";

const paramsInQuery = {
  tags: ["后台用户管理相关接口"],
  summary: "查询单个后台用户信息",
  operationId: "queryUserById",
  parameters: [
    {
      name: "id",
      in: "query",
      required: true,
      description: "用户id",
      example: 1,
      schema: {
        type: "integer",
        format: "int64",
      },
    },
  ],
  responses: {
    "200": {
      description: "OK",
      content: {
        "*/*": {
          schema: {
            required: ["code", "message", "requestId"],
            type: "object",
            properties: {
              code: {
                type: "string",
                description:
                  "请求结果的状态码.如果这个值返回的是200, 那么就代表业务处理成功了",
                example: "200",
              },
              message: {
                type: "string",
                description:
                  "请求结果的描述. 如果业务处理失败，将会描述失败的原因.",
                example: "success",
              },
              data: {
                type: "object",
                properties: {
                  id: {
                    type: "integer",
                    description: "用户id",
                    format: "int64",
                  },
                  email: {
                    type: "string",
                    description: "邮箱",
                  },
                  realName: {
                    type: "string",
                    description: "姓名",
                  },
                  status: {
                    type: "string",
                    description: "用户状态(ACTIVE, INACTIVE)",
                    enum: ["NEW", "ACTIVE", "INACTIVE"],
                  },
                },
                description: "查询用户响应结果",
              },
              requestId: {
                type: "string",
                description: "请求ID",
                example: "00000001",
              },
            },
            description: "HTTP响应的统一结构体",
          },
        },
      },
    },
  },
};

const paramsInBody = {
  tags: ["币种管理相关接口"],
  summary: "分页查询币种信息",
  operationId: "pageSearch_4",
  requestBody: {
    content: {
      "application/json": {
        schema: {
          required: ["count", "pageNum", "pageSize", "params"],
          type: "object",
          properties: {
            pageNum: {
              maximum: 1024,
              minimum: 1,
              type: "integer",
              description: "页码数, 该值的范围是1~1024",
              format: "int32",
              example: 1,
            },
            pageSize: {
              maximum: 99999,
              minimum: 1,
              type: "integer",
              description: "页大小, 该值的范围是1~99999",
              format: "int32",
              example: 1,
            },
            count: {
              type: "boolean",
              description:
                "是否查询总记录数。查询总记录数会严重影响性能，非必要不查询. ",
              example: false,
            },
            params: {
              type: "object",
              properties: {
                nameOrCode: {
                  type: "string",
                  description: "币种代码或者币种名称",
                },
                status: {
                  type: "string",
                  description: "币种状态(NEW, ACTIVE, INACTIVE)",
                  enum: ["NEW", "ACTIVE", "INACTIVE"],
                },
              },
              description: "币种列表查询参数",
            },
            orders: {
              maxItems: 10,
              minItems: 0,
              type: "array",
              description: "分页查询的排序规则, 最大不能超过10个排序字段",
              items: {
                required: ["fieldName", "type"],
                type: "object",
                properties: {
                  fieldName: {
                    type: "string",
                    description: "字段名称",
                  },
                  type: {
                    type: "string",
                    description: "排序规则",
                    enum: ["asc", "desc"],
                  },
                },
                description: "分页查询排序字段结构体",
              },
            },
          },
          description: "分页查询统一结构体",
        },
      },
    },
    required: true,
  },
  responses: {
    "200": {
      description: "OK",
      content: {
        "*/*": {
          schema: {
            required: ["code", "message", "requestId"],
            type: "object",
            properties: {
              code: {
                type: "string",
                description:
                  "请求结果的状态码.如果这个值返回的是200, 那么就代表业务处理成功了",
                example: "200",
              },
              message: {
                type: "string",
                description:
                  "请求结果的描述. 如果业务处理失败，将会描述失败的原因.",
                example: "success",
              },
              data: {
                required: ["result"],
                type: "object",
                properties: {
                  total: {
                    type: "integer",
                    description:
                      "总记录数。如果请求参数中'是否查询总记录数'为true则此参数将会返回.",
                    format: "int64",
                  },
                  result: {
                    type: "array",
                    description: "分页查询的结果",
                    items: {
                      type: "object",
                      properties: {
                        coin: {
                          type: "object",
                          properties: {
                            id: {
                              type: "integer",
                              description: "币种id",
                              format: "int64",
                            },
                            sourceId: {
                              type: "integer",
                              description:
                                "crypto来源Id，crypto_datasource表的Id",
                              format: "int64",
                            },
                            name: {
                              type: "string",
                              description: "Coin的名称，如：Bitcoin",
                            },
                            code: {
                              type: "string",
                              description: "Coin的符号，如：BTC",
                            },
                            status: {
                              type: "string",
                              description: "币种状态(NEW,ACTIVE,INACTIVE)",
                            },
                          },
                          description: "币种管理",
                        },
                        networks: {
                          type: "array",
                          description: "币种支持的网络",
                          items: {
                            type: "object",
                            properties: {
                              id: {
                                type: "integer",
                                description: "表Id，coin唯一标识",
                                format: "int64",
                              },
                              coinId: {
                                type: "integer",
                                description:
                                  "币种的唯一标识，用于将网络关联到币种",
                                format: "int64",
                              },
                            },
                            description: "区块网络返回结果",
                          },
                        },
                      },
                      description: "coin分页查询结果",
                    },
                  },
                },
                description: "分页查询结果统计结构体",
              },
              requestId: {
                type: "string",
                description: "请求ID",
                example: "00000001",
              },
            },
            description: "HTTP响应的统一结构体",
          },
        },
      },
    },
  },
};

const generate = (data, methods: string) => {
  return generateApiRequestCode({
    url: "/users",
    method: methods,
    operationObject: data,
    apiConfig: apiConfigNormalizer({
      swaggerJSONPath: "",
      swaggerUiUrl:
        "http://192.168.104.10:31082/swagger/?urls.primaryName=%E5%90%8E%E5%8F%B0%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3API",
    }),
  });
};

describe("api 生成", function () {
  it("参数在url上", async () => {
    const result = await generate(paramsInQuery, "get");
    expect(result).toMatchSnapshot();
  });

  it("参数在body里", async function () {
    const result = await generate(paramsInBody, "post");
    expect(result).toMatchSnapshot();
  });

  it("参数或响应为空", async () => {
    let parmas = JSON.parse(JSON.stringify(paramsInBody));
    parmas.parameters = [];
    parmas.responses = {};
    const result = await generate(parmas, "post");
    expect(result).toMatchSnapshot();
  });
});
