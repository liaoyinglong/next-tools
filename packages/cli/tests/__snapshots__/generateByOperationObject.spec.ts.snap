// Vitest Snapshot v1

exports[`api 生成 > 参数在body里 1`] = `
"// 这个文件由 @dune2/cli 自动生成，不要手动修改，否则会被覆盖
import { RequestBuilder } from '@dune2/tools';
import requestFn from '@/utils/request';
/**
 * 分页查询币种信息
 * @tags 币种管理相关接口
 * @see http://192.168.104.10:31082/swagger/?urls.primaryName=%E5%90%8E%E5%8F%B0%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3API#/%E5%B8%81%E7%A7%8D%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3%E6%8E%A5%E5%8F%A3/pageSearch_4
 */
export const usersPostApi = new RequestBuilder<usersPostApi.Req, usersPostApi.Res>({
  url: '/users',
  method: 'post',
  requestFn,
});

export namespace usersPostApi {
 /**
 * 分页查询统一结构体
 */
export interface Req {
  /**
   * 页码数, 该值的范围是1~1024
   */
  pageNum: number;
  /**
   * 页大小, 该值的范围是1~99999
   */
  pageSize: number;
  /**
   * 是否查询总记录数。查询总记录数会严重影响性能，非必要不查询.
   */
  count: boolean;
  /**
   * 币种列表查询参数
   */
  params: {
    /**
     * 币种代码或者币种名称
     */
    nameOrCode?: string;
    /**
     * 币种状态(NEW, ACTIVE, INACTIVE)
     */
    status?: \\"NEW\\" | \\"ACTIVE\\" | \\"INACTIVE\\";
  };
  /**
   * 分页查询的排序规则, 最大不能超过10个排序字段
   *
   * @minItems 0
   * @maxItems 10
   */
  orders?: {
    /**
     * 字段名称
     */
    fieldName: string;
    /**
     * 排序规则
     */
    type: \\"asc\\" | \\"desc\\";
  }[];
}

 
 /**
 * 分页查询结果统计结构体
 */
export interface Res {
  /**
   * 总记录数。如果请求参数中'是否查询总记录数'为true则此参数将会返回.
   */
  total?: number;
  /**
   * 分页查询的结果
   */
  result: {
    /**
     * 币种管理
     */
    coin?: {
      /**
       * 币种id
       */
      id?: number;
      /**
       * crypto来源Id，crypto_datasource表的Id
       */
      sourceId?: number;
      /**
       * Coin的名称，如：Bitcoin
       */
      name?: string;
      /**
       * Coin的符号，如：BTC
       */
      code?: string;
      /**
       * 币种状态(NEW,ACTIVE,INACTIVE)
       */
      status?: string;
    };
    /**
     * 币种支持的网络
     */
    networks?: {
      /**
       * 表Id，coin唯一标识
       */
      id?: number;
      /**
       * 币种的唯一标识，用于将网络关联到币种
       */
      coinId?: number;
    }[];
  }[];
}

};"
`;

exports[`api 生成 > 参数在url上 1`] = `
"// 这个文件由 @dune2/cli 自动生成，不要手动修改，否则会被覆盖
import { RequestBuilder } from '@dune2/tools';
import requestFn from '@/utils/request';
/**
 * 查询单个后台用户信息
 * @tags 后台用户管理相关接口
 * @see http://192.168.104.10:31082/swagger/?urls.primaryName=%E5%90%8E%E5%8F%B0%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3API#/%E5%90%8E%E5%8F%B0%E7%94%A8%E6%88%B7%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3%E6%8E%A5%E5%8F%A3/queryUserById
 */
export const usersGetApi = new RequestBuilder<usersGetApi.Req, usersGetApi.Res>({
  url: '/users',
  method: 'get',
  requestFn,
});

export namespace usersGetApi {
 export interface Req {
  /**
   * 用户id
   */
  id: number;
}

 
 /**
 * 查询用户响应结果
 */
export interface Res {
  /**
   * 用户id
   */
  id?: number;
  /**
   * 邮箱
   */
  email?: string;
  /**
   * 姓名
   */
  realName?: string;
  /**
   * 用户状态(ACTIVE, INACTIVE)
   */
  status?: \\"NEW\\" | \\"ACTIVE\\" | \\"INACTIVE\\";
}

};"
`;

exports[`api 生成 > 参数或响应为空 1`] = `
"// 这个文件由 @dune2/cli 自动生成，不要手动修改，否则会被覆盖
import { RequestBuilder } from '@dune2/tools';
import requestFn from '@/utils/request';
/**
 * 分页查询币种信息
 * @tags 币种管理相关接口
 * @see http://192.168.104.10:31082/swagger/?urls.primaryName=%E5%90%8E%E5%8F%B0%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3API#/%E5%B8%81%E7%A7%8D%E7%AE%A1%E7%90%86%E7%9B%B8%E5%85%B3%E6%8E%A5%E5%8F%A3/pageSearch_4
 */
export const usersPostApi = new RequestBuilder<usersPostApi.Req, usersPostApi.Res>({
  url: '/users',
  method: 'post',
  requestFn,
});

export namespace usersPostApi {
 export interface Req {}

 
 export type Res = any;
};"
`;
